File: CodeChef/GCD.java
import java.math.BigInteger;
import java.util.*;
import java.lang.*;
public class Solution {
	public static long MOD = 1000000007;
	public static long conversion (String s){
		long res = 0;
		for (int i = 0; i<s.length(); i++){
			res = res*10;
			res = (res + s.charAt(i) - '0') % MOD;
		}
		
		return res;
	}
	
	public static void main(String args[]) throws Exception {
		Scanner sc = new Scanner(System.in);

		int t = Integer.parseInt(sc.nextLine());
		BigInteger mod = new BigInteger(String.valueOf(1000000007));
		
		while (t-- > 0){
			String s = sc.nextLine();
			String v[] = s.split(" ");
			//BigInteger a = new BigInteger(String.valueOf(conversion (v[0])));
			BigInteger a = new BigInteger(v[0]);
			BigInteger b = new BigInteger(v[1]);
			BigInteger n = new BigInteger(v[2]);
			long x1,x2,x3,ans,x4;
			x3=abs(x1-x2);
			x3=new BigInteger(x3);
			if(a==b){
        	x1=(a.modPow(n, mod));
        	x2=(b.modPow(n, mod));
        	ans=((x1+x2)%mod);
    }
    else{
    x1=(a.modPow(n, x3));
    x2=(b.modPow(n, x3));
    x4=new BigInteger((x1+x2)%x3);
    ans=x4.gcd(x3)%mod;
			System.out.println(ans);
		}
	}
}}
'''a,b,n = input().split()
    a,b,n = int(a),int(b),int(n)
    flag = 0
    m = 10 ** 9 + 7
    if a > b:
        p = a-b
    elif a == b:
        x = pow(a,n,m)
        y = pow(b,n,m)
        print((x+y)%m)
        flag = 1
    else:
        p = b-a
    
    x = pow(a,n,p)
    y = pow(b,n,p)
    if flag == 0:
        print(math.gcd((x+y)%p,p)%m)'''

File: Templates/GCD_LargeNumber/GCD.cpp
#include<bits/stdc++.h>
using namespace std;
#define m 1000000007
long long moduloMultiplication(long long a,long long b,
long long mod)
{
    long long res = 0;
    a %= mod;
    while (b)
    {
        if (b & 1)
            res = (res + a) % mod;
        a = (2 * a) % mod;

        b >>= 1;
    }

    return res;
}

long long int calpow( long long int A,long long  int B, long long int M )
{
	 if(B == 0)
            return 1;
	if(B%2==0)
	    return calpow(moduloMultiplication(A,A,M),B/2,M);
	else
	    return moduloMultiplication(A,calpow(moduloMultiplication(A,A,M),B/2,M),M);

}
long long int find_ans(long long int a,long long int b,long long int n){
    long long int x1=0,x2=0,x3,ans;
    x3=abs(a-b);
    if(a==b){
        x1=calpow(a,n,m);
        x2=calpow(b,n,m);
        return ((x1+x2)%m);
    }
    x1=calpow(a,n,x3);
    x2=calpow(b,n,x3);
    ans=__gcd((x1+x2)%x3,x3)%m;
    return ans;


}
int main(){
long long int a,b,n,t,x1,x2,x3,ans;
cin>>t;

while(t--){

    cin>>a>>b>>n;
    ans=find_ans(a,b,n);
    cout<<ans<<endl;

    }


}



